code/modules/instruments/songs/editor.dm

```diff
@@ -0,0 +1,252 @@
+					Every note in a chord will play together, with chord timed by the tempo.<br>
+					<br>
+					Notes are played by the names of the note, and optionally, the accidental, and/or the octave number.<br>
+					By default, every note is natural and in octave 3. Defining otherwise is remembered for each note.<br>
+					Example: <i>C,D,E,F,G,A,B</i> will play a C major scale.<br>
+					After a note has an accidental placed, it will be remembered: <i>C,C4,C,C3</i> is <i>C3,C4,C4,C3</i><br>
+					Chords can be played simply by seperating each note with a hyphon: <i>A-C#,Cn-E,E-G#,Gn-B</i><br>
+					A pause may be denoted by an empty chord: <i>C,E,,C,G</i><br>
+					To make a chord be a different time, end it with /x, where the chord length will be length<br>
+					defined by tempo / x: <i>C,G/2,E/4</i><br>
+					Combined, an example is: <i>E-E4/4,F#/2,G#/8,B/8,E3-E4/4</i>
+					<br>
+					Lines may be up to [MUSIC_MAXLINECHARS] characters.<br>
+					A song may only contain up to [MUSIC_MAXLINES] lines.<br>
+					"}
+		else
+			dat += "<B><A href='?src=[REF(src)];help=2'>Show Help</A></B><BR>"
+
+	var/datum/browser/popup = new(user, "instrument", parent?.name || "instrument", 700, 500)
+	popup.set_content(dat.Join(""))
+	popup.set_title_image(user.browse_rsc_icon(parent.icon, parent.icon_state))
+	popup.open()
+
+/**
+  * Parses a song the user has input into lines and stores them.
+  */
+/datum/song/proc/ParseSong(text)
+	set waitfor = FALSE
+	//split into lines
+	lines = splittext(text, "\n")
+	if(lines.len)
+		var/bpm_string = "BPM: "
+		if(findtext(lines[1], bpm_string, 1, length(bpm_string) + 1))
+			var/divisor = text2num(copytext(lines[1], length(bpm_string) + 1)) || 120 // default
+			tempo = sanitize_tempo(600 / round(divisor, 1))
+			lines.Cut(1, 2)
+		else
+			tempo = sanitize_tempo(5) // default 120 BPM
+		if(lines.len > MUSIC_MAXLINES)
+			to_chat(usr, "Too many lines!")
+			lines.Cut(MUSIC_MAXLINES + 1)
+		var/linenum = 1
+		for(var/l in lines)
+			if(length_char(l) > MUSIC_MAXLINECHARS)
+				to_chat(usr, "Line [linenum] too long!")
+				lines.Remove(l)
+			else
+				linenum++
+		updateDialog(usr)		// make sure updates when complete
+
+/datum/song/Topic(href, href_list)
+	if(!usr.canUseTopic(parent, TRUE, FALSE, FALSE, FALSE))
+		usr << browse(null, "window=instrument")
+		usr.unset_machine()
+		return
+
+	parent.add_fingerprint(usr)
+
+	if(href_list["newsong"])
+		lines = new()
+		tempo = sanitize_tempo(5) // default 120 BPM
+		name = ""
+
+	else if(href_list["import"])
+		var/t = ""
+		do
+			t = html_encode(input(usr, "Please paste the entire song, formatted:", text("[]", name), t)  as message)
+			if(!in_range(parent, usr))
+				return
+
+			if(length_char(t) >= MUSIC_MAXLINES * MUSIC_MAXLINECHARS)
+				var/cont = input(usr, "Your message is too long! Would you like to continue editing it?", "", "yes") in list("yes", "no")
+				if(cont == "no")
+					break
+		while(length_char(t) > MUSIC_MAXLINES * MUSIC_MAXLINECHARS)
+		ParseSong(t)
```

Comment:
Is this tabbed correctly?

Reply:
yep

Reply:
Weird, on github it looks like the `while` is empty.